README.md
# Commandy


Commandy is simple Web framework written in Haskell

- define your domain commands as types
- define your domain commnads results as types
- define all your fields with smart constructor validation
- implicit mapping of json objects to command types (via template haskell mapper)
- auto generation of all errors or success result
- easy to change internal transport (websocket used)
- easy to change command data representation (json, grpc, row text, etc)


Here's an example:

```haskell

data RegArgs = RegArgs
  { username :: Maybe UserName,
    email :: Email
  }
  deriving (Show)

data RegError
  = UserExists UserName
  | EmailUsed Email
  | OtherRegError String
  deriving (Generic, Show)

reg :: RegArgs -> IO (ErrorOrResult RegError ResultOk_)
reg (args@RegArgs {username = username, email = email}) = do
  let uname = fromMaybe "NULL" (show <$> username)
  let qargs = [show email, "16", uname]
  conn <- connect defaultConnectInfo {connectPassword = "root", connectDatabase = "test"}
  do
    r <- (try $ execute conn ("insert into user (email,age,name) values (?,?,?)") qargs) :: IO (Either MySQLError Int64)
    case r of
      Left e -> if errNumber e == 1062 then return . DError $ EmailUsed email else error "something WRONG!"
      Right _ -> do uid <- insertID conn; print $ "uid" ++ show uid; return $ DResult ROK_

```




This is draft for now!


CHANGELOG.md
# Revision history for commandy

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/Commandy.hs
{-# LANGUAGE OverloadedStrings #-}
module Commandy where

import Data.Maybe
import Data.Validation

-- Field required err or actual fields errors
data FieldRequired a = FieldRequired | FieldErrors a deriving (Show)

data FieldType = FNull | FRequired deriving (Show,Eq)

data Field e r = Field {
  name :: String,
  ftype :: FieldType,
  smc :: String -> Validation e r
  --default ??
}

-- check that required field has error or not
e :: Maybe (Either a b) -> Bool
e Nothing = True
e (Just (Left _)) = True
e _ = False

-- check that nullable field has error or not
e_ :: Maybe (Either a b) -> Bool
e_ Nothing = False
e_ (Just (Left _)) = True
e_ _ = False

-- extract error of required field
errs :: Monoid a => Maybe (Either a b) -> FieldRequired a
errs (Just (Left a)) = FieldErrors a
errs Nothing = FieldRequired
errs _ = FieldErrors mempty

-- errs _ = error "use only to extract errors (not succes results)"

-- extract error of an null field
errs_ :: Monoid a => Maybe (Either a b) -> a
errs_ (Just (Left a)) = a
errs_ _ = mempty

-- errs_ _ = error "use only to extract errors (not succes results)"

-- extract result of req field
res :: Maybe (Either a b) -> b
res (Just (Right b)) = b
res _ = error "use only to extract succes results"

-- extract result of null field
res_ :: Maybe (Either a b) -> Maybe b
res_ (Just (Right b)) = Just b
res_ Nothing = Nothing
res_ _ = error "use only to extract succes results"




app/WS_Server.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}

module Main where

import Commandy
import Control.Exception
import Control.Lens hiding ((.=))
import Control.Monad
import Control.Monad (forever)
import Data.Aeson
import qualified Data.Aeson.Key as LB
import Data.Aeson.Types
import qualified Data.ByteString.Lazy as LB
import qualified Data.ByteString.Lazy.UTF8 as LBU8
import Data.Commandy.Json
import Data.Commandy.Types
import qualified Data.Commandy.Types as CT
import Data.Either (fromLeft, fromRight, isLeft)
import qualified Data.Either.Combinators as EithQ
import Data.Maybe
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import Data.Text.Lazy.Encoding
import GHC.Generics
import qualified Network.WebSockets as WS
import Users.Args
import Users.Json
import Users.Mapper
import Users.Types

jsonStringAct :: LB.ByteString
jsonStringAct = "{ \"action\": \"reg\", \"username\": \"John\" , \"email\": \"john@gmail.com\" }"

jsonStringAct2 = "{ \"action\": \"reg\", \"username\": \"John\" , \"email\": \"john!gmail.com\" }" :: LB.ByteString

exF' :: FromJSON b => LB.ByteString -> Key -> Either String b
exF' t n = do
  result <- eitherDecode t -- "{\"name\":\"Dave\",\"age\":2}"
  flip parseEither result $ \obj -> obj .: n

-- !!! APPEND ROUTES HERE - TEMPLATE HASKELL GENERATION !!!
$(srouter "regRouter" ''RegArgs 'reg)
$(srouter "loginRouter" ''LoginArgs 'login)

-- TODO : implement template haskell routing with empty args !

-- $(srouter "userlist" ''UsersListArgs 'userlist)

router :: LB.ByteString -> IO LB.ByteString
router b = case act of
  -- this routing can be done via template haskell generation.. need it ?
  Right "reg" -> regRouter b
  Right "login" -> loginRouter b
  Right act -> do
    putStrLn "recieved uncknown act!"
    return $ "uncknown act " `mappend` LBU8.fromString act
  Left err -> do
    putStrLn "recieved bad json!"
    return $ "error: " `mappend` LBU8.fromString err
  where
    (act :: Either String String) = exF' b "action"

chandle :: WS.Connection -> IO ()
chandle conn = forever $ do
  msg <- WS.receiveData conn
  -- catch any exception here
  -- https://stackoverflow.com/questions/6009384/exception-handling-in-haskell
  res <- try $ router msg :: IO (Either SomeException LB.ByteString)
  case res of
    Left ex -> do
      let excinf = "Caught exception: " ++ show ex
      putStrLn $ excinf
      WS.sendTextData conn $ LBU8.fromString excinf
    Right resp ->
      do
        putStrLn "start send to client..."
        WS.sendTextData conn resp
        putStrLn "done."

-- WS.sendTextData conn $ msg `T.append` ", handle"
-- WS.sendTextData conn resp

main =
  print "serving 8080"
    >> ( WS.runServer "127.0.0.1" 8080 $ \pending -> do
           conn <- WS.acceptRequest pending
           print "accept conn"
           chandle conn
       )

-- ws://127.0.0.1:8080
-- { "action": "reg", "username": "John" , "email": "john@gmail.com" }
-- https://codereview.stackexchange.com/questions/200310/using-mysql-with-wai-and-warp
-- https://hackage.haskell.org/package/resource-pool-0.3.1.0/docs/Data-Pool.html

-- https://begriffs.com/posts/2014-10-25-creating-package-hackage.html
-- https://cabal.readthedocs.io/en/3.4/developing-packages.html

-- logging
-- env vars https://hackage.haskell.org/package/load-env
-- polling
-- ddos
-- primitive types

app/Users/Mapper.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TemplateHaskell #-}

module Users.Mapper where

-- need to print params! ??

import qualified Commandy as CD
import Data.Aeson
import qualified Data.Aeson.Key as AK
import qualified Data.ByteString.Lazy as LB
import qualified Data.ByteString.Lazy.Char8 as CHLB
import qualified Data.Commandy.Json as CJS
import qualified Data.Commandy.Types as CT
import Data.Either
import Data.List
import qualified Data.Map as M
import Data.Maybe
import Data.Text (pack, splitOn, unpack)
import Debug.Trace
import Language.Haskell.TH
import Language.Haskell.TH.Quote
import Language.Haskell.TH.Syntax as THS
import qualified TH.ReifySimple as RS
import Users.Types

-- TODO check real maybe
ismaybe :: Type -> Bool
ismaybe (AppT (ConT _) _) = True
ismaybe (ConT _) = False
ismaybe (_) = error "unexpected types configuration: use Maybe a | a"

extr (AppT (ConT _) t) = t
extr t@(ConT _) = t
extr _ = error "unexpected type conf in EXTR"

tname (ConT n) = show n
tname (_) = error "unexpected type cons"

-- TODO: this explicit mapping can be removed, all info containt inside typeclass instances of fromstring
-- TODO search for instances fromString and build mapping from fieldtype => errorFtype
tMapping =
  M.fromList
    [ ("Data.Commandy.Types.Email", ConT $ mkName "Data.Commandy.Types.EmailValidationError"),
      ("Data.Commandy.Types.UserName", ConT $ mkName "Data.Commandy.Types.UserNameValidationError"),
      ("Data.Commandy.Types.Password", ConT $ mkName "Data.Commandy.Types.PasswordValidationError")
    ]

-- eex_ = [e| \d k -> ((fromRight undefined) . CT.fromString) <$> (exF d k) |]
-- eex = [e| \d k -> fromJust $ ((fromRight undefined) . CT.fromString) <$> (exF d k) |]

x_Errs :: Type -> Q Exp
x_Errs ft = [e|\d k -> errs $ ((CT.fromString)) ::$ (ft) <$> (exF d k)|]

x_Errs_ :: Type -> Q Exp
x_Errs_ ft = [e|\d k -> errs_ $ ((CT.fromString)) ::$ (ft) <$> (exF d k)|]

srouter :: String -> Name -> Name -> Q [Dec]
srouter name r dact = do
  -- https://hackage.haskell.org/package/template-haskell-2.19.0.0/docs/src/Language.Haskell.TH.Syntax.html#TyConI
  tr@(TyConI (DataD _ n _ _ [constr@(RecC cn varbangtype)] _)) <- reify r
  let args_and_types = [(unpack $ last $ splitOn "." $ pack . show $ n, extr tn, ismaybe tn) | (n, _, tn) <- varbangtype]

  eex_ <- [e|\d k -> ((fromRight undefined) . CT.fromString) <$> (exF d k)|]
  eex <- [e|\d k -> fromJust $ ((fromRight undefined) . CT.fromString) <$> (exF d k)|]

  exv <- [e|\d k -> ((CT.fromString)) <$> (exF d k)|]

  d <-
    [d|
      -- rt :: LB.ByteString -> Either Value RegArgs
      rt :: LB.ByteString -> IO LB.ByteString
      rt rdata =
        do
          
          -- lookupTypeName ??
          
          -- putStrLn $ "injected!" ++ (show args_and_types) ++ "\n" ++  "\n" ++ (show dact_)
          
          -- extract fields values from jsom
          -- apply SmartConstructors to fields
          -- check any errors
          -- if no arrors build args
          -- if error build errors as json value and encode it          
          -- make AST, wrap it into Q, splice with $ and bind name args to constructed code

          let args =
                $( let ce = ConE $ cn
                       barg fns ismb = (AppE (AppE (if ismb then eex_ else eex) (VarE 'rdata)) (LitE $ StringL fns))
                    in return $ foldl (AppE) ce [barg fns ismb | (fns, _, ismb) <- args_and_types]
                 )

          let is_verrors =
                $( let ff = VarE '(||)
                       barg fns = (AppE (AppE exv (VarE 'rdata)) (LitE $ StringL fns))
                       dsig_ e ft = SigE e $ AppT (ConT ''Maybe) (AppT (AppT (ConT ''Either) (AppT ListT ((tMapping M.! (tname ft))))) ft)
                       chose_e ft = if ismaybe ft then (VarE 'CD.e_) else (VarE 'CD.e)
                    in return $ foldl (AppE) ff [(AppE (chose_e ft) $ dsig_ (barg fns) ft) | (fns, ft, ismb) <- args_and_types]
                 )

          -- putStrLn $ show is_verrors

          let errors_value =
                $( let -- build json value (toJson)
                       -- buildt ft ismb = if ismb then (AppT ListT ((tMapping M.! (tname ft)))) else AppT (ConT ''CD.FieldRequired) (AppT ListT ((tMapping M.! (tname ft))))
                       -- berrs fns ismb ft = SigE (AppE (AppE (if ismb then ex_errs_ else ex_errs ) (VarE 'rdata) ) (LitE $ StringL fns)) $ buildt ft ismb

                       barg fns = (AppE (AppE exv (VarE 'rdata)) (LitE $ StringL fns))
                       dsig_ e ft = SigE e $ AppT (ConT ''Maybe) (AppT (AppT (ConT ''Either) (AppT ListT ((tMapping M.! (tname ft))))) ft)
                       chose_e ft = if ismaybe ft then (VarE 'CD.errs_) else (VarE 'CD.errs)
                    in -- errs fn ismb rdata = AppE (VarE 'toJSON) (SigE (LitE $ StringL "some errors") $ ConT ''String)
                       return $ AppE (VarE 'object) $ ListE [AppE (AppE (VarE '(.=)) (LitE (StringL fn))) (AppE (chose_e ft) $ dsig_ (barg fn) ft) | (fn, ft, ismb) <- args_and_types]
                 )
          -- let erros_value = toJSON $ ("123" :: String)
          let wrapped_errors_value =
                object
                  [ "result" .= ("fail" :: String),
                    "data" .= errors_value
                  ]

          -- if not is_verrors then router_ (Right args) $(return $ (VarE dact)) else return $ encode wrapped_errors_value
          if not is_verrors
            then do
              putStrLn "Action validation ok. Start domain action..."
              (encode <$> ($(return (VarE dact)) args)) >>= \r -> print "Domain action done!" >> return r
            else return $ encode wrapped_errors_value

            -- {-# LANGUAGE OverloadedStrings #-}
            -- :set -XOverloadedStrings
            -- object :: [Pair] -> Value

            -- customValue :: Value
            -- customValue = object
            --   [ "list_price" .= (150000 :: Int)
            --   , "sale_price" .= (143000 :: Int)
            --   , "description" .= ("2-bedroom townhouse" :: String)
            --   ]
      |]

  let funName = mkName name
      [SigD _ funSig, FunD _ funBody] = d
      d' = [SigD funName funSig, FunD funName funBody]
  return d'

-- hasErrors :: LB.ByteString -> [(Key, FieldType, SC)] -> Bool
-- hasErrors t fields = any id [if ft == FNull then e_ $ pf t fn sc else e $ pf t fn sc | (fn, ft, (SC sc)) <- fields]

-- if is_errors -- hasErrors t regFields
--   then
--     Left $
--       RegArgsErrors
--         (gErrs_ t $ _username regFields_) -- (errs_ username_)
--         (gErrs t $ _email regFields_) -- (errs email_)
--   else
--     Right $
--       RegArgs
--         (gRes_ t $ _username regFields_) -- (res_ username_)
--         (gRes t $ _email regFields_) -- (res email_)
-- where
--   -- is_errors = e username_ || e_ email_
--   is_errors = (isFerr t $ _username regFields_) || (isFerr t $ _email regFields_)

-- https://hackage.haskell.org/package/modulespection

-- make or in AST: VarE $ mkName $ "GHC.Classes.||"
-- make type in AST: ConT $ mkName "GHC.Maybe.Maybe"
-- ConT $ ''Either

app/Users/Types.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE OverloadedStrings #-}

module Users.Types where

import Commandy
import Control.Exception
import Data.Commandy.Types
import Data.Int (Int64)
import Data.Maybe
import Database.MySQL.Base (MySQLError (errNumber))
import Database.MySQL.Simple
import Database.MySQL.Simple.QueryResults
import Database.MySQL.Simple.Result
import GHC.Generics

-- best loacted not here? req for core.. - this errors out of domain (more like parsing errs)

data RegArgsErrors = RegArgsErrors
  { username_error :: [UserNameValidationError],
    email_error :: FieldRequired [EmailValidationError]
  }
  deriving (Generic, Show)

data RegArgs = RegArgs
  { username :: Maybe UserName,
    email :: Email
  }
  deriving (Show)

data RegError
  = UserExists UserName
  | EmailUsed Email
  | OtherRegError String
  deriving (Generic, Show)

reg :: RegArgs -> IO (ErrorOrResult RegError ResultOk_)
reg (args@RegArgs {username = username, email = email}) = do
  -- error "catch me!"
  print args
  let uname = fromMaybe "NULL" (show <$> username)
  let qargs = [show email, "16", uname]
  conn <- connect defaultConnectInfo {connectPassword = "root", connectDatabase = "test"}

  do
    r <- (try $ execute conn ("insert into user (email,age,name) values (?,?,?)") qargs) :: IO (Either MySQLError Int64)
    case r of
      Left e -> if errNumber e == 1062 then return . DError $ EmailUsed email else error "something WRONG!"
      Right _ -> do uid <- insertID conn; print $ "uid" ++ show uid; return $ DResult ROK_

data LoginArgs = LoginArgs
  { email :: Email,
    pwd :: Password
  }
  deriving (Show)

data LoginArgsErrors
  = UserBanned Email
  | OtherLoginError String
  deriving (Generic, Show)

login :: LoginArgs -> IO (ErrorOrResult LoginArgsErrors ResultOk_)
login args = do
  putStrLn "implement login please!"
  return $ DResult ROK_

data NoErrors = NoErrors deriving (Generic, Show)

data UsersListArgs = UsersListArgs deriving (Generic, Show)

userlist :: UsersListArgs -> IO (ErrorOrResult NoErrors ResultOk_)
userlist args = do
  putStrLn "implement show all users please!"
  return $ DResult ROK_